gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsUtils.java

```diff
@@ -0,0 +1,178 @@
+      return "true".equals(useClientCertificate);
+    }
+
+    @Override
+    public String useMtlsEndpoint() {
+      return envProvider.getenv(GOOGLE_API_USE_MTLS_ENDPOINT);
+    }
+
+    @Override
+    public String getKeyStorePassword() {
+      return "";
+    }
+
+    @Override
+    public KeyStore getKeyStore() throws IOException, GeneralSecurityException {
+      if (keyStore != null) {
+        // return the cached key store
+        return keyStore;
+      }
+      try {
+        // Load the cert provider command from the json file.
+        InputStream stream = new FileInputStream(metadataPath);
+        List<String> command = extractCertificateProviderCommand(stream);
+
+        // Run the command and timeout after 1000 milliseconds.
+        Process process = new ProcessBuilder(command).start();
+        int exitCode = runCertificateProviderCommand(process, 1000);
+        if (exitCode != 0) {
+          throw new IOException("Cert provider command failed with exit code: " + exitCode);
+        }
+
+        // Create mTLS key store with the input certificates from shell command.
+        keyStore = SecurityUtils.createMtlsKeyStore(process.getInputStream());
+        return keyStore;
+      } catch (FileNotFoundException ignored) {
+        // file doesn't exist
+        return null;
+      } catch (InterruptedException e) {
+        throw new IOException("Interrupted executing certificate provider command", e);
+      }
+    }
+
+    @VisibleForTesting
+    static List<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+        throws IOException {
+      JsonParser parser = new JacksonFactory().createJsonParser(contextAwareMetadata);
+      ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+      return json.getCommands();
+    }
+
+    @VisibleForTesting
+    static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+        throws IOException, InterruptedException {
+      long startTime = System.currentTimeMillis();
+      long remainTime = timeoutMilliseconds;
+      boolean terminated = false;
+
+      do {
+        try {
+          // Check if process is terminated by polling the exitValue, which throws
+          // IllegalThreadStateException if not terminated.
+          commandProcess.exitValue();
+          terminated = true;
+          break;
+        } catch (IllegalThreadStateException ex) {
+          if (remainTime > 0) {
+            Thread.sleep(Math.min(remainTime + 1, 100));
+          }
+        }
+        remainTime = remainTime - (System.currentTimeMillis() - startTime);
+      } while (remainTime > 0);
+
+      if (!terminated) {
+        commandProcess.destroy();
+        throw new IOException("cert provider command timed out");
+      }
+
+      return commandProcess.exitValue();
+    }
+  }
+
+  private static final MtlsProvider MTLS_PROVIDER = new DefaultMtlsProvider();
```

Comment:
Minor: please  move the static constant declaration at the top of the class implementation.

Reply:
No longer have this static constant.